### PR TITLE
APEXCORE-683 Apex client should support application packages on HDFS

### PR DESCRIPTION
**Pull request by vrozov**

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 1]
```diff
@@ -460,6 +461,23 @@ protected void restoreOutput(PrintStream originalStream)
     }
   }
 
+  AppPackage newAppPackageInstance(URI uri) throws IOException, ZipException
```
Comment (ajaygit158-zz) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: ZipException is not thrown by any of the methods used, can be removed.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 2]
```diff
@@ -2842,12 +2860,20 @@ public void execute(String[] args, ConsoleReader reader) throws Exception
...
-          AppPackage ap = new AppPackage(new File(filename));
-          ap.close();
+          if (scheme == null || scheme.equals("file")) {
+            filename = expandFileName(uri.getPath(), true);
+            new AppPackage(new FileInputStream(new File(filename)), true).close();
+          } else {
+            try (FileSystem fs = FileSystem.newInstance(uri, conf)) {
+              new AppPackage(fs.open(new Path(filename)), true).close();
+            }
+          }
```
Comment (ajaygit158-zz) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Instead of newAppPackageInstance having similar code, can we have AppPackage(URI, boolean).

File: engine/src/main/java/com/datatorrent/stram/client/AppPackage.java [hunk 3]
```diff
@@ -255,19 +285,51 @@ private void processAppProperties()
...
+
+    JarEntry entry = input.getNextJarEntry();
+    while (entry != null) {
+      File newFile = new File(directory, entry.getName());
+      if (entry.isDirectory()) {
+        newFile.mkdirs();
+      } else {
+        FileOutputStream fileoutputstream = new FileOutputStream(newFile);
+        IOUtils.copy(input, fileoutputstream);
+        fileoutputstream.close();
```
Comment (ajaygit158-zz) [inline engine/src/main/java/com/datatorrent/stram/client/AppPackage.java]: We can do this in try() - finally block to ensure outputstream is always closed.

**Review by ajaygit158-zz — COMMENTED** [review]
@vrozov Added a few comments.

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java (hunk 2 above)
Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: I do not understand the proposal. What is the meaning of the boolean flag? close?

Comment (ajaygit158-zz) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Sorry, I assumed the boolean parameter passed to fs.open() is getting passed to AppPackage(). My proposal was having AppPackage(URI uri)

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 4]
```diff
@@ -2842,18 +2861,15 @@ public void execute(String[] args, ConsoleReader reader) throws Exception
       }
 
       if (commandLineInfo.args.length > 0) {
-        String filename = expandFileName(commandLineInfo.args[0], true);
-
         // see if the first argument is actually an app package
-        try {
-          AppPackage ap = new AppPackage(new File(filename));
-          ap.close();
+        try (AppPackage ap = newAppPackageInstance(new URI(commandLineInfo.args[0]), false)) {
```
Comment (PramodSSImmaneni) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java]: Why not expand file name before passing it to open like it was being done before?

File: engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java [hunk 5]
```diff
@@ -460,7 +460,26 @@ protected void restoreOutput(PrintStream originalStream)
...
 
-  AppPackage newAppPackageInstance(File f) throws IOException, ZipException
+  AppPackage newAppPackageInstance(URI uri, boolean suppressOutput) throws IOException
+  {
+    PrintStream outputStream = suppressOutput ? suppressOutput() : null;
+    try {
+      final String scheme = uri.getScheme();
+      if (scheme == null || scheme.equals("file")) {
+        return new AppPackage(new FileInputStream(new File(expandFileName(uri.getPath(), true))), true);
+      } else {
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:470]: Why not do special handling for hdfs only when scheme is "dfs type" such as hdfs or s3 types. In other cases go with default url handlers using uri.toURL().openStream()? This will be able to handle other sources such as http and ftp.

Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:470]: It is not handled in other places as well. The goal of the PR is to add support for files on hdfs only. Support for other schemes may be done as a separate PR.

Comment (pramodin) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:470]: Can we check that scheme before doing hdfs specific handling?

Comment (vrozov) [reply engine/src/main/java/com/datatorrent/stram/cli/ApexCli.java:470]: It may be safer to delegate to the FileSystem to deal with schemes it supports and throw an Exception for those that it does not support. In the future, if support for "http" or "ftp" and other schemes are added, I'd recommend to handle them similar to "file", by checking for those schemes explicitly.

**Comment by pramodin** [comment]
Can you add some description in the JIRA as to where this will be useful for users. The current description is very simple and not informative for users.

**Comment by vrozov** [comment]
OK, done. It is natural for the native yarn application like Apex to support HDFS and other DFS in addition to local file system.

File: engine/src/test/java/com/datatorrent/stram/client/AppPackageTest.java [hunk 6]
```diff
@@ -57,24 +55,17 @@
   String appPackageDir = "src/test/resources/testAppPackage/mydtapp";
 
   @BeforeClass
-  public static void starting()
+  public static void starting() throws IOException, JSONException
   {
+    File file = StramTestSupport.createAppPackageFile();
```
Comment (pramodin) [inline engine/src/test/java/com/datatorrent/stram/client/AppPackageTest.java:60]: Why the change move it out of the try?

Comment (vrozov) [reply engine/src/test/java/com/datatorrent/stram/client/AppPackageTest.java:60]: if an exception is thrown before application package is created, there is no need to close it.
